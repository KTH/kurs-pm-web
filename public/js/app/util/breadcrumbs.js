import React from 'react'
import { Breadcrumbs } from '@kth/kth-reactstrap/dist/components/utbildningsinfo'
import { createRoot } from 'react-dom/client'

export function renderBreadcrumbsIntoKthHeader(courseCode, languageAbbr) {
  const breadcrumbContainer = document.getElementById('breadcrumbs-header')
  if (breadcrumbContainer) {
    const root = createRoot(breadcrumbContainer)
    root.render(<Breadcrumbs include={'aboutCourse'} courseCode={courseCode} language={languageAbbr} />)
  }
}
