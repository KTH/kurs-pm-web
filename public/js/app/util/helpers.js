import i18n from '../../../../i18n'

export const seasonStr = (translate, semesterCode = '') =>
  `${translate.season[semesterCode.toString()[4]]}${semesterCode.toString().slice(0, 4)}`

export const aboutCourseStr = (translate, courseCode = '') => `${translate.aboutCourse} ${courseCode}`

export const getDateFormat = (date, language) => {
  if (language === 'Svenska' || language === 1 || language === 'sv' || date.length === 0) {
    return date
  }

  var parts = date.split('/')
  var dt = new Date(parseInt(parts[2], 10), parseInt(parts[1], 10) - 1, parseInt(parts[0], 10))
  const options = {
    day: 'numeric',
    month: 'short',
    year: 'numeric',
  }

  return dt.toLocaleDateString('en-GB', options)
}

export const concatMemoName = (semester, applicationCodes = [], langAbbr = 'sv') => {
  const langIndex = langAbbr === 'en' ? 0 : 1
  const { memoLabel } = i18n.messages[langIndex].messages
  if (applicationCodes.length === 0) {
    return `${memoLabel} ${seasonStr(i18n.messages[langIndex].extraInfo, semester)}-${''}`
  }
  return `${memoLabel} ${seasonStr(i18n.messages[langIndex].extraInfo, semester)}-${
    applicationCodes.length > 1 ? `${applicationCodes[0]}...` : applicationCodes[0]
  }`
}

export const concatMemoNameShort = (semester, applicationCodes = [], langAbbr = 'sv') => {
  const langIndex = langAbbr === 'en' ? 0 : 1
  return `${seasonStr(i18n.messages[langIndex].extraInfo, semester)}-${applicationCodes.map(code => code).join('-')}`
}

export const memoNameWithoutApplicationCode = (semester, langAbbr = 'sv') => {
  const langIndex = langAbbr === 'en' ? 0 : 1
  const { memoLabel } = i18n.messages[langIndex].messages
  return `${memoLabel} ${seasonStr(i18n.messages[langIndex].extraInfo, semester)}`
}

export const memoNameWithCourseCode = (courseCode, semester, applicationCodes = [], langAbbr = 'sv') => {
  const langIndex = langAbbr === 'en' ? 0 : 1
  const { memoLabel } = i18n.messages[langIndex].messages

  return `${memoLabel} ${courseCode} ${seasonStr(i18n.messages[langIndex].extraInfo, semester)}-${
    applicationCodes.length > 1 ? `${applicationCodes[0]}...` : applicationCodes[0]
  }`
}
export const removeDuplicates = elements => {
  return elements.filter((term, index) => elements.indexOf(term) === index)
}
export const roundShortNameWithStartdate = (round, langAbbr = 'sv') => {
  const langIndex = langAbbr === 'en' ? 0 : 1
  const { startdate } = i18n.messages[langIndex].aboutMemoLabels
  const roundStartDate = getDateFormat(round.firstTuitionDate, langIndex)
  const pattern = /[a-zA-Z]\w*\s\d{4}[-]\d{1}/
  const { memoName: memoNames } = round

  if ('memoName' in round) return `${memoNames} (${startdate} ${roundStartDate})`

  if ('shortName' in round && round.shortName !== '') {
    let shortMemoNames = ''
    if (pattern.test(round.shortName)) {
      shortMemoNames = `${seasonStr(i18n.messages[langIndex].extraInfo, round.term || round.semester)}`
    } else {
      shortMemoNames = round.shortName.replace(/ m.fl./g, '')
    }
    return `${shortMemoNames} (${startdate} ${roundStartDate})`
  }
  return `${seasonStr(
    i18n.messages[langIndex].extraInfo,
    round.term || round.semester
  )} (${startdate} ${roundStartDate})`
}

export const doubleSortOnAnArrayOfObjects = (arr, par1, par2, langIndex) => {
  const arrWithSortedMemoNames = [...arr].map(round => {
    const regEx = /\s*\(.*?\)\s*/g
    const pattern = /[a-zA-Z]\w*\s\d{4}[-]\d{1}/
    if ('memoName' in round) {
      const memoNames = round.memoName.replace(regEx, '').replace(/ m.fl./g, '')
      const memoNamesArray = memoNames.split(', ').map(item => {
        if (pattern.test(item)) {
          return `${seasonStr(i18n.messages[langIndex].extraInfo, round.term || round.semester)}`
        }
        return item
      })

      const uniqueMemoNamesArray = removeDuplicates(memoNamesArray)
      const sortedMemoNamesArray = uniqueMemoNamesArray.sort()

      return {
        ...round,
        memoName: sortedMemoNamesArray.join(', '),
        shortName: sortedMemoNamesArray[0],
      }
    }

    return round
  })

  let sortedArray = arrWithSortedMemoNames.sort((a, b) => {
    if (Number(a[par1]) == Number(b[par1])) {
      if (a[par2] === b[par2]) return 0
      return a[par2] < b[par2] ? -1 : 1
    } else {
      return Number(a[par1]) < Number(b[par1]) ? -1 : 1
    }
  })

  return sortedArray
}

export const memoNameWithoutCourseCode = (semester, applicationCodes = [], langAbbr = 'sv') => {
  const langIndex = langAbbr === 'en' ? 0 : 1
  const { memoLabel, prepositionFor } = i18n.messages[langIndex].messages
  return `${memoLabel} ${prepositionFor} ${seasonStr(
    i18n.messages[langIndex].extraInfo,
    semester
  )}-${applicationCodes.join('-')}`
}

export const formatCredits = (credits, creditUnitAbbr, language) => {
  const localeCredits = language === 'sv' ? credits.toLocaleString('sv-SE') : credits.toLocaleString('en-GB')
  const creditUnit = language === 'sv' ? creditUnitAbbr : 'credits'
  return `${localeCredits} ${creditUnit}`
}
