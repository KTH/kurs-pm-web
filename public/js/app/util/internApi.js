import axios from 'axios'

// eslint-disable-next-line consistent-return
async function getTermsWithCourseRounds(courseCode) {
  try {
    const result = await axios.get(`/termsWithCourseRounds/${courseCode}`)
    if (result) {
      if (result.status >= 400) {
        return 'ERROR-INTERN-API' + result.status
      }
      const { data } = result
      return data
    }
  } catch (error) {
    if (error.response) {
      throw new Error('Unexpected error from intern API ' + error.message)
    }
    throw error
  }
}

export default getTermsWithCourseRounds
