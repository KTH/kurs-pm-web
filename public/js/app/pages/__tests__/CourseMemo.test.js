import React from 'react'
// import { useWebContext } from '../context/WebContext'
import { BrowserRouter, Routes, Route } from 'react-router-dom'

import { render, screen } from '@testing-library/react'
import { WebContextProvider } from '../../context/WebContext'

import '@testing-library/jest-dom/extend-expect'

import CourseMemo, { redirectToAbout } from '../CourseMemo'
import { formatCredits } from '../../util/helpers'

const { getAllByRole } = screen

describe('Redirect function in <CourseMemo>', () => {
  test('return data for redirect for the link from the personal meny', () => {
    const locationUrlAboutMemo = { pathname: '/MF2140/om-kurs-pm' }
    const data = redirectToAbout('MF2140', locationUrlAboutMemo)
    expect(data).toBeNull()
  })
  test('return data for redirect for the link from the personal meny', () => {
    const locationUrlFromPersonalMenu = { pathname: '/MF2140/20222/1' }
    const { noMemoData, semester, roundIds } = redirectToAbout('MF2140', locationUrlFromPersonalMenu)
    expect(noMemoData).toBeTruthy()
    expect(semester).toBe('20222')
    expect(roundIds).toStrictEqual(['1'])
  })
  test('return data for redirect for the link with the memos endpoint which is missing', () => {
    const locationUrlWithMissingMemoEndpoint = { pathname: '/SF1624/SF162420231-1' }
    const { noMemoData, semester, roundIds } = redirectToAbout('SF1624', locationUrlWithMissingMemoEndpoint)
    expect(noMemoData).toBeTruthy()
    expect(semester).toBe('20231')
    expect(roundIds).toStrictEqual(['1'])
  })
  test('return data with two rounds for redirect for the link with memo endpoint which is missing', () => {
    const locationUrlWithMissingMemoEndpoint = { pathname: '/SF1624/SF162420232-5-6' }
    const { noMemoData, semester, roundIds } = redirectToAbout('SF1624', locationUrlWithMissingMemoEndpoint)
    expect(noMemoData).toBeTruthy()
    expect(semester).toBe('20232')
    expect(roundIds).toStrictEqual(['5', '6'])
  })

  test('return data for redirect for the full link from the personal meny', () => {
    const locationUrlFromPersonalMenu = { pathname: '/kurs-pm/MF2140/20222/1' }
    const { noMemoData, semester, roundIds } = redirectToAbout('MF2140', locationUrlFromPersonalMenu)
    expect(noMemoData).toBeTruthy()
    expect(semester).toBe('20222')
    expect(roundIds).toStrictEqual(['1'])
  })

  test('return data for redirect for the link with the memos endpoint which is missing', () => {
    const locationUrlWithMissingMemoEndpoint = { pathname: '/kurs-pm/SF1624/SF162420231-1' }
    const { noMemoData, semester, roundIds } = redirectToAbout('SF1624', locationUrlWithMissingMemoEndpoint)
    expect(noMemoData).toBeTruthy()
    expect(semester).toBe('20231')
    expect(roundIds).toStrictEqual(['1'])
  })
  test('return data with two rounds for redirect for the link with memo endpoint which is missing', () => {
    const locationUrlWithMissingMemoEndpoint = { pathname: '/kurs-pm/SF1624/SF162420232-5-6' }
    const { noMemoData, semester, roundIds } = redirectToAbout('SF1624', locationUrlWithMissingMemoEndpoint)
    expect(noMemoData).toBeTruthy()
    expect(semester).toBe('20232')
    expect(roundIds).toStrictEqual(['5', '6'])
  })
})

// {"pathname":"/MF2140/om-kurs-pm","search":"","hash":"","state":{"noMemoData":true,"semester":"20222","roundIds":["1"]},"key":"1wefohoy"}
describe('Component <CourseMemo>', () => {
  test('renders a course memo', () => {
    const context = {
      browserConfig: { imageStorageUri: 'localhost://' },
      memoData: {
        courseTitle: '',
        visibleInMemo: {},
        syllabusValid: { validFromTerm: '20181' },
      },
      memoDatas: [
        {
          semester: '',
          applicationCodes: [],
          memoCommonLangAbbr: '',
          syllabusValid: { validFromTerm: '20181' },
        },
      ],
      userLanguageIndex: 0,
      credits: 7.5,
      syllabusValid: { validFromTerm: '20181' },
    }
    render(
      <BrowserRouter basename={'/'} location={{}}>
        <WebContextProvider configIn={context}>
          <Routes>
            <Route path="/" element={<CourseMemo />} />
          </Routes>
        </WebContextProvider>
      </BrowserRouter>
    )
  })
})

const spring181 = {
  courseCode: 'TEST121',
  memoEndPoint: '12345',
  semester: '20181',
  applicationCodes: [12345],
  memoCommonLangAbbr: 'en',
  credits: 7.5,
  lastChangeDate: 'Fri Jan 27 2018 12:04:37 GMT+0000 (Coordinated Universal Time)',
  version: 2,
  syllabusValid: { validFromTerm: '20181' },
}
const spring1823 = {
  courseCode: 'TEST121',
  memoEndPoint: '34567',
  semester: '20181',
  applicationCodes: [23456, 34567],
  memoCommonLangAbbr: 'en',
  credits: 7.5,
  lastChangeDate: 'Fri Dec 1 2017 12:04:37 GMT+0000 (Coordinated Universal Time)',
  version: 6,
  syllabusValid: { validFromTerm: '20181' },
}
const spring184 = {
  courseCode: 'TEST121',
  memoEndPoint: '45678',
  semester: '20181',
  applicationCodes: [45678],
  memoCommonLangAbbr: 'en',
  credits: 7.5,
  syllabusValid: { validFromTerm: '20181' },
}
const spring1924 = {
  courseCode: 'TEST121',
  memoEndPoint: '34567',
  semester: '20191',
  applicationCodes: [23456, 34567],
  memoCommonLangAbbr: 'en',
  credits: 7.5,
  syllabusValid: { validFromTerm: '20191' },
}
const spring193 = {
  courseCode: 'TEST121',
  memoEndPoint: '34567',
  semester: '20191',
  applicationCodes: [34567],
  memoCommonLangAbbr: 'en',
  credits: 7.5,
  syllabusValid: { validFromTerm: '20191' },
}
const fall190 = {
  courseCode: 'TEST121',
  memoEndPoint: '190',
  semester: '20192',
  applicationCodes: [],
  memoCommonLangAbbr: 'en',
  credits: 7.5,
  syllabusValid: { validFromTerm: '20191' },
}

describe('Page CourseMemo', () => {
  beforeAll(() => {
    jest.useFakeTimers('modern')
    jest.setSystemTime(new Date(2020, 1, 1))
  })

  afterAll(() => {
    jest.useRealTimers()
  })

  test('renders a side menu and a page structure', done => {
    const memoDatas = [spring181, spring1823, spring184, spring1924, spring193, fall190]
    const context = {
      browserConfig: { imageStorageUri: 'localhost://' },
      memoData: { ...spring181, courseTitle: '', visibleInMemo: {} },
      memoDatas,
      memoLanguageIndex: 0,
      userLanguageIndex: 0,
      credits: 7.5,
      activeMemoEndPoint: id => id === '12345',
      roundIds: [2, 3],
      applicationCodes: ['12345'],
      courseCode: 'TEST121',
    }

    const { asFragment } = render(
      <BrowserRouter basename={'/'} location={{}}>
        <WebContextProvider configIn={context}>
          <Routes>
            <Route path="/" element={<CourseMemo />} />
          </Routes>
        </WebContextProvider>
      </BrowserRouter>
    )
    expect(asFragment()).toMatchSnapshot()

    done()
  })

  test('renders a side menu links', done => {
    const memoDatas = [spring181, spring1823, spring184, spring1924, spring193, fall190]
    const context = {
      browserConfig: { imageStorageUri: 'localhost://' },
      memoData: { ...spring181, courseTitle: '', visibleInMemo: {} },
      memoDatas,
      memoLanguageIndex: 0,
      userLanguageIndex: 0,
      credits: 7.5,
      activeMemoEndPoint: id => id === '12345',
      roundIds: [],
      courseCode: 'TEST121',
      applicationCodes: [],
    }

    const { asFragment } = render(
      <BrowserRouter basename={'/'} location={{}}>
        <WebContextProvider configIn={context}>
          <Routes>
            <Route path="/" element={<CourseMemo />} />
          </Routes>
        </WebContextProvider>
      </BrowserRouter>
    )

    const links = getAllByRole('link', { name: /Course memo.*\d*-\d*/i })
    expect(links.length).toEqual(4)
    const expectedLinks = [
      ['Course memo Spring 2018-12345', '/kurs-pm/TEST121/12345'],
      ['Course memo Spring 2018-23456...', '/kurs-pm/TEST121/34567'],
      ['Course memo Spring 2018-45678', '/kurs-pm/TEST121/45678'],
      ['Course memo Autumn 2019-undefined', '/kurs-pm/TEST121/190'],
    ]

    links.forEach((l, index) => {
      expect(l.getAttribute('aria-label')).toBe(expectedLinks[index][0])
      expect(l.textContent).toBe(expectedLinks[index][0])
      expect(l.href).toBe(`http://localhost${expectedLinks[index][1]}`)
    })

    done()
  })
})
