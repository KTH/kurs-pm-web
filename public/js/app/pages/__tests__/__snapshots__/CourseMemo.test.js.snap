// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page CourseMemo renders a side menu and a page structure 1`] = `
<DocumentFragment>
  <section
    class="cover-page d-none d-print-block"
  >
    <h1>
      TEST121  7.5 credits
    </h1>
    <p
      class="memo-subtitle"
    >
      Kurs-PM VT 2018-1
    </p>
    <p>
      Version 2 – null
    </p>
    <h2>
      Kursomgång
    </h2>
    <p>
      <span>
        <br />
      </span>
    </p>
    <h2>
      Undervisningsspråk
    </h2>
    <p />
    <h2>
      Kursen ges av
    </h2>
    <p />
    <ul
      class="link-list cover-page-links"
    >
      <li>
        * Information hämtas från N/A
      </li>
      <li>
        Utskrift från sidan: 
        <a
          href="http://localhost/"
        >
          http://localhost/
        </a>
      </li>
    </ul>
  </section>
  <div
    class="p-print-4  row"
  >
    <nav
      aria-labelledby="kth-local-navigation-title--mobile"
      class="kth-local-navigation--mobile"
    >
      <button
        class="kth-button menu"
        id="kth-local-navigation-title--mobile"
      >
        <span>
          About course TEST121
        </span>
      </button>
      <dialog
        class="kth-mobile-menu left"
      >
        <div
          class="kth-mobile-menu__navigation"
        >
          <button
            class="kth-icon-button close"
          >
            <span
              class="kth-visually-hidden"
            >
              Close
            </span>
          </button>
        </div>
        <div
          class="mobile-menu__content"
        >
          <a
            class="kth-button back"
          >
            Course and programme directory
          </a>
          <h2>
            About course TEST121
          </h2>
          <ul>
            <li>
              <a
                href="/student/kurser/kurs/TEST121"
              >
                 Before course selection
              </a>
            </li>
            <li
              class="kth-local-navigation__submenu"
            >
              <h3>
                Prepare and take course
              </h3>
              <ul>
                <li>
                  <a
                    href="/kurs-pm/TEST121/om-kurs-pm"
                  >
                    Course memo
                  </a>
                </li>
                <li>
                  <a
                    href="/kurs-pm/TEST121/181"
                  >
                    Course memo Spring 2018-1
                  </a>
                </li>
                <li>
                  <a
                    href="/kurs-pm/TEST121/1823"
                  >
                    Course memo Spring 2018-2...
                  </a>
                </li>
                <li>
                  <a
                    href="/kurs-pm/TEST121/184"
                  >
                    Course memo Spring 2018-4
                  </a>
                </li>
                <li>
                  <a
                    href="/kurs-pm/TEST121/1924"
                  >
                    Course memo Spring 2019-2...
                  </a>
                </li>
                <li>
                  <a
                    href="/kurs-pm/TEST121/193"
                  >
                    Course memo Spring 2019-3
                  </a>
                </li>
                <li>
                  <a
                    href="/kurs-pm/TEST121/190"
                  >
                    Course memo Autumn 2019-
                  </a>
                </li>
              </ul>
            </li>
            <li>
              <a
                href="/kursutveckling/TEST121"
              >
                Course development
              </a>
            </li>
            <li>
              <a
                href="/kursutveckling/TEST121/arkiv"
              >
                Archive
              </a>
            </li>
          </ul>
        </div>
      </dialog>
    </nav>
    <nav
      aria-labelledby="local-navigation-title"
      class="kth-local-navigation col"
      id="mainMenu"
    >
      <a
        class="kth-button back"
      >
        Course and programme directory
      </a>
      <h2
        id="local-navigation-title"
      >
        About course TEST121
      </h2>
      <ul>
        <li>
          <a
            href="/student/kurser/kurs/TEST121"
          >
             Before course selection
          </a>
        </li>
        <li
          class="kth-local-navigation__submenu"
        >
          <h3>
            Prepare and take course
          </h3>
          <ul>
            <li>
              <a
                href="/kurs-pm/TEST121/om-kurs-pm"
              >
                Course memo
              </a>
            </li>
            <li>
              <a
                href="/kurs-pm/TEST121/181"
              >
                Course memo Spring 2018-1
              </a>
            </li>
            <li>
              <a
                href="/kurs-pm/TEST121/1823"
              >
                Course memo Spring 2018-2...
              </a>
            </li>
            <li>
              <a
                href="/kurs-pm/TEST121/184"
              >
                Course memo Spring 2018-4
              </a>
            </li>
            <li>
              <a
                href="/kurs-pm/TEST121/1924"
              >
                Course memo Spring 2019-2...
              </a>
            </li>
            <li>
              <a
                href="/kurs-pm/TEST121/193"
              >
                Course memo Spring 2019-3
              </a>
            </li>
            <li>
              <a
                href="/kurs-pm/TEST121/190"
              >
                Course memo Autumn 2019-
              </a>
            </li>
          </ul>
        </li>
        <li>
          <a
            href="/kursutveckling/TEST121"
          >
            Course development
          </a>
        </li>
        <li>
          <a
            href="/kursutveckling/TEST121/arkiv"
          >
            Archive
          </a>
        </li>
      </ul>
    </nav>
    <main
      class="col col-print-12"
      id="mainContent"
    >
      <header>
        <h1
          aria-labelledby="page-heading page-sub-heading"
          id="page-heading"
        >
          Kurs-PM VT 2018
        </h1>
        <div
          id="page-sub-heading-wrapper"
        >
          <p
            aria-hidden="true"
            id="page-sub-heading"
          >
            TEST121  7.5 credits
          </p>
          <p
            class="d-print-none d-none d-sm-block"
            id="page-sub-heading-admin-link"
          >
            <a
              href="/kursinfoadmin/kurser/kurs/TEST121?l=undefined"
              title="Administer About course"
            >
              Administer About course
            </a>
          </p>
        </div>
      </header>
      <div
        class="row"
      >
        <div
          class="text-break col-print-12 content-center col-lg-8"
          id="flexible-content-of-center"
        >
          <p
            class="mb-4"
          >
            Rubriker markerade med en asterisk ( * ) kommer från kursplan version VT 2018
          </p>
          <section
            aria-labelledby="contentAndOutcomes"
            class="section-wrapper"
          >
            <h2
              id="contentAndOutcomes"
            >
              Innehåll och lärandemål
            </h2>
            <article
              aria-labelledby="courseContent"
              id="contentAndOutcomes-courseContent"
            >
              <h3
                id="courseContent"
              >
                Kursinnehåll
                <sup>
                  <svg
                    class="syllabus-marker-icon"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M478.21 334.093L336 256l142.21-78.093c11.795-6.477 15.961-21.384 9.232-33.037l-19.48-33.741c-6.728-11.653-21.72-15.499-33.227-8.523L296 186.718l3.475-162.204C299.763 11.061 288.937 0 275.48 0h-38.96c-13.456 0-24.283 11.061-23.994 24.514L216 186.718 77.265 102.607c-11.506-6.976-26.499-3.13-33.227 8.523l-19.48 33.741c-6.728 11.653-2.562 26.56 9.233 33.037L176 256 33.79 334.093c-11.795 6.477-15.961 21.384-9.232 33.037l19.48 33.741c6.728 11.653 21.721 15.499 33.227 8.523L216 325.282l-3.475 162.204C212.237 500.939 223.064 512 236.52 512h38.961c13.456 0 24.283-11.061 23.995-24.514L296 325.282l138.735 84.111c11.506 6.976 26.499 3.13 33.227-8.523l19.48-33.741c6.728-11.653 2.563-26.559-9.232-33.036z"
                    />
                  </svg>
                </sup>
              </h3>
              <span>
                Ingen information tillagd
              </span>
            </article>
            <article
              aria-labelledby="learningOutcomes"
              id="contentAndOutcomes-learningOutcomes"
            >
              <h3
                id="learningOutcomes"
              >
                Lärandemål
                <sup>
                  <svg
                    class="syllabus-marker-icon"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M478.21 334.093L336 256l142.21-78.093c11.795-6.477 15.961-21.384 9.232-33.037l-19.48-33.741c-6.728-11.653-21.72-15.499-33.227-8.523L296 186.718l3.475-162.204C299.763 11.061 288.937 0 275.48 0h-38.96c-13.456 0-24.283 11.061-23.994 24.514L216 186.718 77.265 102.607c-11.506-6.976-26.499-3.13-33.227 8.523l-19.48 33.741c-6.728 11.653-2.562 26.56 9.233 33.037L176 256 33.79 334.093c-11.795 6.477-15.961 21.384-9.232 33.037l19.48 33.741c6.728 11.653 21.721 15.499 33.227 8.523L216 325.282l-3.475 162.204C212.237 500.939 223.064 512 236.52 512h38.961c13.456 0 24.283-11.061 23.995-24.514L296 325.282l138.735 84.111c11.506 6.976 26.499 3.13 33.227-8.523l19.48-33.741c6.728-11.653 2.563-26.559-9.232-33.036z"
                    />
                  </svg>
                </sup>
              </h3>
              <span>
                Ingen information tillagd
              </span>
            </article>
          </section>
          <section
            aria-labelledby="prep"
            class="section-wrapper"
          >
            <h2
              id="prep"
            >
              Förberedelser inför kursstart
            </h2>
            <article
              aria-labelledby="literature"
              id="prep-literature"
            >
              <h3
                id="literature"
              >
                Kurslitteratur
              </h3>
              <span>
                Ingen information tillagd
              </span>
            </article>
            <article
              aria-labelledby="permanentDisability"
              id="prep-permanentDisability"
            >
              <h3
                id="permanentDisability"
              >
                Stöd för studenter med funktionsnedsättning
              </h3>
              <span>
                Ingen information tillagd
              </span>
            </article>
          </section>
          <section
            aria-labelledby="reqToFinal"
            class="section-wrapper"
          >
            <h2
              id="reqToFinal"
            >
              Examination och slutförande
            </h2>
            <article
              aria-labelledby="gradingScale"
              id="reqToFinal-gradingScale"
            >
              <h3
                id="gradingScale"
              >
                Betygsskala
                <sup>
                  <svg
                    class="syllabus-marker-icon"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M478.21 334.093L336 256l142.21-78.093c11.795-6.477 15.961-21.384 9.232-33.037l-19.48-33.741c-6.728-11.653-21.72-15.499-33.227-8.523L296 186.718l3.475-162.204C299.763 11.061 288.937 0 275.48 0h-38.96c-13.456 0-24.283 11.061-23.994 24.514L216 186.718 77.265 102.607c-11.506-6.976-26.499-3.13-33.227 8.523l-19.48 33.741c-6.728 11.653-2.562 26.56 9.233 33.037L176 256 33.79 334.093c-11.795 6.477-15.961 21.384-9.232 33.037l19.48 33.741c6.728 11.653 21.721 15.499 33.227 8.523L216 325.282l-3.475 162.204C212.237 500.939 223.064 512 236.52 512h38.961c13.456 0 24.283-11.061 23.995-24.514L296 325.282l138.735 84.111c11.506 6.976 26.499 3.13 33.227-8.523l19.48-33.741c6.728-11.653 2.563-26.559-9.232-33.036z"
                    />
                  </svg>
                </sup>
              </h3>
              <span>
                Ingen information tillagd
              </span>
            </article>
            <article
              aria-labelledby="examination"
              id="reqToFinal-examination"
            >
              <h3
                id="examination"
              >
                Examination
                <sup>
                  <svg
                    class="syllabus-marker-icon"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M478.21 334.093L336 256l142.21-78.093c11.795-6.477 15.961-21.384 9.232-33.037l-19.48-33.741c-6.728-11.653-21.72-15.499-33.227-8.523L296 186.718l3.475-162.204C299.763 11.061 288.937 0 275.48 0h-38.96c-13.456 0-24.283 11.061-23.994 24.514L216 186.718 77.265 102.607c-11.506-6.976-26.499-3.13-33.227 8.523l-19.48 33.741c-6.728 11.653-2.562 26.56 9.233 33.037L176 256 33.79 334.093c-11.795 6.477-15.961 21.384-9.232 33.037l19.48 33.741c6.728 11.653 21.721 15.499 33.227 8.523L216 325.282l-3.475 162.204C212.237 500.939 223.064 512 236.52 512h38.961c13.456 0 24.283-11.061 23.995-24.514L296 325.282l138.735 84.111c11.506 6.976 26.499 3.13 33.227-8.523l19.48-33.741c6.728-11.653 2.563-26.559-9.232-33.036z"
                    />
                  </svg>
                </sup>
              </h3>
              <span>
                Ingen information tillagd
              </span>
            </article>
            <article
              aria-labelledby="ethicalApproach"
              id="reqToFinal-ethicalApproach"
            >
              <h3
                id="ethicalApproach"
              >
                Etiskt förhållningssätt
                <sup>
                  <svg
                    class="syllabus-marker-icon"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M478.21 334.093L336 256l142.21-78.093c11.795-6.477 15.961-21.384 9.232-33.037l-19.48-33.741c-6.728-11.653-21.72-15.499-33.227-8.523L296 186.718l3.475-162.204C299.763 11.061 288.937 0 275.48 0h-38.96c-13.456 0-24.283 11.061-23.994 24.514L216 186.718 77.265 102.607c-11.506-6.976-26.499-3.13-33.227 8.523l-19.48 33.741c-6.728 11.653-2.562 26.56 9.233 33.037L176 256 33.79 334.093c-11.795 6.477-15.961 21.384-9.232 33.037l19.48 33.741c6.728 11.653 21.721 15.499 33.227 8.523L216 325.282l-3.475 162.204C212.237 500.939 223.064 512 236.52 512h38.961c13.456 0 24.283-11.061 23.995-24.514L296 325.282l138.735 84.111c11.506 6.976 26.499 3.13 33.227-8.523l19.48-33.741c6.728-11.653 2.563-26.559-9.232-33.036z"
                    />
                  </svg>
                </sup>
              </h3>
              <span>
                Ingen information tillagd
              </span>
            </article>
          </section>
          <section
            aria-labelledby="extra"
            class="section-wrapper"
          >
            <h2
              id="extra"
            >
              Ytterligare Information
            </h2>
            <p>
              <i>
                Ingen information tillagd
              </i>
            </p>
          </section>
          <section
            aria-labelledby="print-memo-contacts"
            class="d-none d-print-block print-contacts"
          >
            <h2
              class="info"
              id="print-memo-contacts"
            >
              Kontakter
            </h2>
            <div
              class="text-break"
            >
              <h3>
                Kursansvarig
              </h3>
              <p />
              <h3>
                Lärare
              </h3>
              <p />
              <h3>
                Examinator
              </h3>
              <p />
            </div>
          </section>
        </div>
        <div
          class="d-print-none content-right col-lg-4"
        >
          <section
            aria-labelledby="memo-facts"
          >
            <h2
              class="d-none"
              id="memo-facts"
            >
              Fakta om kursomgång
            </h2>
            <div
              class="info-box text-break"
            >
              <h3>
                Startdatum
              </h3>
              <p>
                Obligatoriskt innehåll saknas
              </p>
              <h3>
                Kursomgång
              </h3>
              <p>
                Obligatoriskt innehåll saknas
              </p>
              <h3>
                Undervisningsspråk
              </h3>
              <p>
                Obligatoriskt innehåll saknas
              </p>
              <h3>
                Kursen ges av
              </h3>
              <p>
                Obligatoriskt innehåll saknas
              </p>
            </div>
          </section>
          <aside
            aria-labelledby="memo-documents"
          >
            <h2
              class="d-none"
              id="memo-documents"
            >
              Dokument
            </h2>
            <div
              class="info-box text-break"
            >
              <h3>
                Version kurs-PM
              </h3>
              <p>
                Obligatoriskt innehåll saknas
              </p>
              <h3>
                Skriv ut eller spara
              </h3>
              <button
                class="print-link"
                color="link"
                id="print-link"
              >
                Skriv ut eller spara kurs-PM
              </button>
            </div>
          </aside>
          <aside
            aria-labelledby="student-links"
            class="info-box text-break"
          >
            <h2
              id="student-links"
            >
              Relaterad information
            </h2>
            <p>
              <a
                href="https://www.kth.se/student/studier/rattigheter-och-skyldigheter/rattigheter-och-skyldigheter-1.1148520"
                id="link-rightsAndResponsibilities"
                title="Rättigheter och skyldigheter"
              >
                Rättigheter och skyldigheter
              </a>
            </p>
            <p>
              <a
                href="https://www.kth.se/student/studier/kurs"
                id="link-courseAndExamination"
                title="Kurs och examination"
              >
                Kurs och examination
              </a>
            </p>
            <p>
              <a
                href="https://www.kth.se/student/studier/administrera/administrera-dina-studier-1.1143032"
                id="link-administrateYouStudy"
                title="Administrera dina studier"
              >
                Administrera dina studier
              </a>
            </p>
          </aside>
          <section
            aria-labelledby="memo-contacts"
            id="last-element-which-determines-styles"
          >
            <h2
              id="memo-contacts"
            >
              Kontakter
            </h2>
            <div
              class="info-box text-break"
            >
              <h3>
                Kursansvarig
              </h3>
              <p />
              <h3>
                Lärare
              </h3>
              <p />
              <h3>
                Examinator
              </h3>
              <p />
            </div>
          </section>
        </div>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
